Sync with community code in prep for new build structure

fossell opened the pull request:
This PR contains cherry-picked commits from our DTC_post branch that include:
- Bug fixes from our recent bug fix release version v4.0.1
- Adding some community control files to the parm/ directory
- Bug fix suggestions from recent merge with develop.

Merging these mods to develop will sync up the community and emc source codes and allow us to make the final switch and hand off for the new build structure.  
Comments on specific code sections can be found in files changed.

fossell committed:
Fix allocation issue for PSFC in Block 2

fossell committed:
Fix mods for selecting satellite channels in CALRAD_WCLOUD_newcrtm

fossell committed:
Fix grib codes in RQSTFLD.f for wind shear variables

fossell committed:
Fix incorrect EL level for TAUX/TAUY calculation

fossell committed:
Fix typo in Wind Energy Calculation (UPP ID 411)

fossell committed:
Fix index for SEVIRI

fossell committed:
Remove extraneous greater than sign

fossell committed:
Revert MXLVL back to 70 for short term sync of code

fossell committed:
Comment out #if 0 blocks, this syntax fails with some compilers

fossell committed:
Correct pname for max updraft helicity

fossell committed:
Add community control files

fossell commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5137,52 +5137,52 @@ SUBROUTINE CLDRAD
-         endif
-      ENDIF
-#endif
+!#if 0
```
Why is this "#if 0" syntax used?  Does the developer not want this code chunk compiled/used?  This syntax fails with non-intel compilers.  Commented out for testing with community platforms, assuming that was the developer's intent.  Need different handling of this if it should be conditional in some cases.

WenMeng-NOAA replied:
I think these are incidentally added by the UPP developer for testing chemistry fields. We may remove the line "#if 0" and "#endif" and keep the rest of lines from 5141 to 5184.

fossell replied:
I uncommented that code and pushed these mods to my branch

fossell commented on sorc/ncep_post.fd/RQSTFLD.f:
```diff
@@ -2078,44 +2078,44 @@ module RQSTFLD_mod
-     &                      /1,'INST AER FN SMASS   ',164,001,     &
-                             'instn aer fine mode sfc mass con  '/
-#endif
+!#if 0
```
Why is this "#if 0" syntax used?  Does the developer not want this code chunk compiled/used?  This syntax fails with non-intel compilers.  Commented out for testing with community platforms, assuming that was the developer's intent.  Need different handling of this if it should be conditional in some cases.

mkavulich replied:
Kate, I was confused about this as well. It seems to be instituted as part of commit 2cea446a, which is a merge from the post_chem1 branch on vlab which was not transitioned over to github. 

I am not sure why it is failing, as it is valid code per the CPP documentation: https://www.cprogramming.com/reference/preprocessor/if.html

However, I would say that it is bad practice, just as commenting whole code blocks without good reason is bad practice. Without a good reason for them staying in, those sections of code should be removed, not just commented out.

WenMeng-NOAA replied:
This session was added by an UPP developer. She intended to define new chemistry fields. Thanks for catching this. I think we may remove the line  "#if 0" and "#endif", but keep rest of lines from 2082 to 2117.

fossell replied:
I uncommented and removed the directive and pushed the new mods.

fossell commented on sorc/ncep_post.fd/RQSTFLD.f:
```diff
@@ -24,7 +24,7 @@ module RQSTFLD_mod
 !
 !     increase MXFLD each time you add a new field
-      INTEGER, PARAMETER :: MXFLD=950,MXLVL=500
+      INTEGER, PARAMETER :: MXFLD=950,MXLVL=70
```
Per previous discussion, leave this as original 70 for now if we can.  Increasing this forces community to add more zeros to *.parm files for grib1 support, not ideal.   Goal is to keep codes in sync until we finalize the build restructuring and then freeze for community release.

WenMeng-NOAA replied:
We changed MXLV=500 due to gfs v16 has 127 levels.

fossell replied:
Ok, would you be open to us wrapping this with a compiler directive "ifdef COMMCODE" and changing it from a .f suffix to a .F suffix.  We can remove these and change back to .f if desired after the next comm release code freeze?  
The final option is that we modify out .parm files with extra zeros, and also need to modify READCNTRL.f with correct formatting for reading those files.  This is doable, just less appealing from a user standpoint.

WenMeng-NOAA replied:
We may try that option.

fossell replied:
Is there concern about disconnecting history associated with RQSTFLD.f if we move that file to be RQSTFLD.F?  I think git can still follow the content change in the log using the --follow option with git log.  But we would have a couple months with a different file/suffix.

WenMeng-NOAA replied:
I think the short time history disconnection would be fine.

fossell replied:
Moved RQSTFLD.f and READCNTRL.f to .F files and added compiler directives to set mxlvl to 70 just for community build.

fossell committed:
Uncomment new chem code fields. Add compiler directive for commmunity set mxlvls to 07 for grib1

fossell committed:
Move RQSTFLD and READCNTRL to .F files to be preprocessed for COMMCODE compiler directive

WenMeng-NOAA commented on sorc/ncep_post.fd/RQSTFLD.f:
```diff
@@ -2077,8 +2082,6 @@ module RQSTFLD_mod
 !                             'dust cr sfc mass con              '/
       DATA IFILV(686),AVBL(686),IQ(686),IS(686),AVBLGRB2(686)      &
      &                      /1,'DUST25 SFC MASS CON ',161,001,     &
-                             'dust pm25 sfc mass con            '/
```
This line should be added in.

fossell replied:
Good catch, sorry about that.  I'll push this back in.

WenMeng-NOAA commented:
Kate,

Can you sync the recent changes from develop at original repos. so that I can run regression tests.
Thanks,

WenMeng-NOAA commented on parm/post_avblflds.xml:
```diff
@@ -3665,7 +3665,7 @@
        <longname>maximum Updraft Helicity on Specified Height Level Above Ground</longname>
        <pdstmpl>tmpl4_8</pdstmpl>
-       <pname>UPHL</pname>
+       <pname>MXUPHL</pname>
```
This field is currently used by operational HRRR and regional FV3. It is better to keep unchanged.

fossell replied:
In the post_avblflds_raphrrr.xml file, they use MXUPHL for pname of UPP ID 420, so we had changed to match this for community wrf applications.  But if that has changed for newer operational HRRR and region FV3, then I can take this mod out of the PR and we will keep it in our seperate post_avblflds_comupp.xml.

WenMeng-NOAA replied:
Kate,

You are right. I checked fv3sar_post_avblflds.xml and post_avblflds_raphrrr.xml, ID 420 has 
pname MXUPHL. Since gfs currently does not output this field. I think it is fine for changing
UPHL into MXUPHL. Thanks!

fossell committed:
Add line accidently deleted for dust25

fossell committed:
Remove READCNTRL.f and RQSTFLD.f since they were moved to .F files for preprocessing

fossell committed:
Merge remote-tracking branch 'upstream/develop' into dtc_sync

fossell commented:
Wen,
I addressed your comments and pushed the requested mods.  I did not modify the post_avblflds.xml MXUPHL file yet, but did comment, waiting for your response and final guidance. 
I merged the top of develop into this PR as well so you can run regression tests.

WenMeng-NOAA commented:
Kate,
I conducted the UPP regression tests on Dell. The results look good for me. No changes were found. 
We will have a UPP branch develop commit next week for RAP/HRRR implementation.  I will schedule
 your branch merging after it. You might manually pick up some changes from RAP/HRRR 
implementation in RQSTFLD.F.

fossell commented:
Wen,
Great news, thanks!  We will look for RAP/HRRR merge and take those changes into our DTC_post branch when available.

fossell committed:
Manually add ifdef commcode for mxlvl to RQSTFLD.F. Unable to retain during merge

WenMeng-NOAA commented:
Conducted regression tests on Dell, Cray and Hera. No changed results were found. This pull request is ready for merging.
